### DISCUSSION TITLE
Not static, guild specific hybrid commands

### DISCUSSION THREAD

**Discussion by Kadethra** [Rapptz/discord.py#8263] [Q&A] [answered]
Hey,
So I would like to make commands customisable for users, so they can decide which one they have on their guild and which not. I am using hybrid commands which makes a bit more difficult. I tried using the guilds parameter in bot.add_cog, but it doesn't work with hybrids so I opened a feature request. I asked if it's possible to add that hybrid commands, or at least the app_command part of them can be affected bythe guilds parameter and got the answer, that one can already do it with decorators (app_commands.guilds).
Now my question: Since user can change the options, so it's not static, how can I apply it?

Hope my problem is understandable so far.

**Comment by LostLuma** [comment, marked as answer]
You can call [`CommandTree.add_command`](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.CommandTree.add_command) / [`CommandTree.remove_command`](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.CommandTree.remove_command) and then sync the tree for the guild(s) you added / removed the command to at any point, allowing you to change where commands are available on the fly.

This means to make your command available dynamically you can go multiple routes:

The `@guilds` decorator can be used without passing in any guilds, allowing you to specify which guilds the command is for later before syncing:

```py
@commands.hybrid_command()
@app_commands.guilds()
async def example(context: commands.Context) -> None:
   ...

# Somewhere before syncing the tree
bot.tree.add_command(example.app_command, guilds=...)
```

You can also simply register your command as a global command, but before you sync the tree you remove it and add it for the guilds it should be in:

```py
@commands.hybrid_command()
async def example(context: commands.Context) -> None:
   ...

# Somewhere before syncing the tree
bot.tree.remove_command('example')
bot.tree.add_command(example.app_command, guilds=...)
```

**Reply by Kadethra** [reply]
Thx for your help ^^ I am usually using cogs, is there also a way to do it?

**Reply by LostLuma** [reply]
In cogs there's the `cog_load` async setup function in which you can fetch the list of guilds from your database. I can give a more detailed answer later if needed.

**Reply by Kadethra** [reply]
It would be nice if u could explain it more detailed. I don't really know how I could do it.

**Reply by Kadethra** [reply]
Ok nvm I tried to combine the things u said:
```py
async def cog_load(self):
        self.bot.tree.remove_command(module)
        self.bot.tree.add_command(self.app_command, guilds=self.bot.settings.get(module), override=True)
```
It does work but since every app commands is two times on my guild, I thought they would be still there.

Edit: I messed smth up. It works because it is a GroupCog in this case.

**Reply by Kadethra** [reply]
Uhm could you help me to get it to work with a Cog? I really don't get it to work. 
```py
async def cog_load(self):
    guilds = self.bot.settings.get(module)
    for cmd in self.walk_commands():
        # Filter all subcommands out
        # if cmd.parent:
        #     continue
        self.bot.tree.remove_command(cmd.name)
        self.bot.tree.add_command(cmd.app_command, guilds=guilds, override=True)
```
This is everything I currently have (not working). With walk_commands I wanted to get all commands, as if I would do it outside of the function.  override is enabled because otherwise I would get an aleady registered error.

**Reply by LostLuma** [reply]
Could you explain in what way it is not working (post errors if you get any)? I have written a small cog to test and this seems to work just fine:

```py

class Example(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot: commands.Bot = bot

    @commands.hybrid_command()
    async def example(self, context: commands.Context) -> None:
        await context.send('Hello!')

    async def cog_load(self) -> None:
        guilds: List[discord.abc.Snowflake] = [discord.Object(380295555039100932)]

        for command in self.walk_commands():
            if isinstance(command, commands.HybridCommand):
                self.bot.tree.remove_command(command.name)
                self.bot.tree.add_command(command.app_command, guilds=guilds)
```

**Reply by Kadethra** [reply]
I have no idea why but it does not work for me. When I use your exact code I do not get an AlreadyRegistered error anymore but I still see (Edit: and execute) it on the guild which is not in the array. I also just reinstalled discord.py to check if my version is too old (which it should not be) and it did not change anything. I also do not get any errors.

**Reply by SebbyLaw** [reply]
If you want to remove commands from a guild you're going to need to sync a tree to it without those commands.

**Reply by Kadethra** [reply]
> If you want to remove commands from a guild you're going to need to sync a tree to it without those commands.

You mean because they are multiple times on my guild?

**Reply by LostLuma** [reply]
If you see the same command two times that means you've previously synced it as a global command as well as as a guild command. Since your goal is to have guild commands only for these you will need to ensure you no longer add the global command to the tree and call `CommandTree.sync` without the guild parameter once, deleting the global variant of it.

**Reply by Kadethra** [reply]
> If you see the same command two times that means you've previously synced it as a global command as well as as a guild command. Since your goal is to have guild commands only for these you will need to ensure you no longer add the global command to the tree and call `CommandTree.sync` without the guild parameter once, deleting the global variant of it.

Ok fixed it. Earlier I had a testing guild defined and the commands from there never got deleted. Just synced the guild with no commands loaded and now they are gone (They didn't work anymore).

Yes the testing guild is the one I now want to exclude but anyways, after finally removing all unused commands, it still does not work. However I always run the sync parameter globally. Should I change it and basically put a list of every guild in it? idk

**Reply by LostLuma** [reply]
If you use guild commands you need to sync with the guild(s) they are on, yes. Note however that you only need to do this once per guild whenever the set of commands changes, not every time you start your bot.

**Reply by Kadethra** [reply]
Ok so now I tried the following code for syncing (I'll add a check if the commands changed etc. later)
```py
for guild in self.settings.get("all_guilds"):
    self.tree.copy_global_to(guild=guild)
    await self.tree.sync(guild=guild)
```
Additionally I deleted all global commands before, so no duplicates. The example command however still works.

**Reply by LostLuma** [reply]
So your issue is resolved now if I understand correctly?

**Reply by Kadethra** [reply]
No ^^" The example command is still on the guild on which it shouldn't be. I'm sorry for all the troubles.

**Reply by LostLuma** [reply]
If you're still using the code from [your earlier comment](https://github.com/Rapptz/discord.py/discussions/8263#discussioncomment-3213521) that just means the example command "module" is enabled on that guild, or did that change?

I'm not quite sure what the code we're working with here looks like to help you further, maybe link your bot repo or create a [secret gist](https://gist.github.com) with all the relevant files and link it here.

**Reply by Kadethra** [reply]
I currently don't have a repository, since I first wanted to get the main parts done. I would actually like to show the code so I could directly answer any questions about it or even more important to directly try your solutions/ideas.
However if you prefer a gist, I can upload a shortened version of the source code where I deleted every unnecessary elements for this problem (for example the folder structure for loading the modules, logging etc.) so you can run it yourself.

**Reply by LostLuma** [reply]
The location or folder structure doesn't really matter to me I would just like to be able to read the relevant parts at once.

**Reply by Kadethra** [reply]
Here. I shortened it pretty much and it's still not working for me. You should be able to run this version if u want to.
https://gist.github.com/SuchtiZockerLP/06c77e59b84a83ce9a73520d88fa7952

**Reply by LostLuma** [reply]
Thanks. Your code copies the same list of *all* guild IDs as the settings for every extension instead of the modified one as you can [see here](https://gist.github.com/SuchtiZockerLP/06c77e59b84a83ce9a73520d88fa7952#file-utils-py-L44). You will want to use `perms.update({ext: guilds})` instead and then it will work.

**Reply by Kadethra** [reply]
Sry rewrite error because it was another function. Doesn't work even with the change.

**Reply by LostLuma** [reply]
I see now. I assumed the library will add app commands from cogs to the tree before calling the `cog_load` function, but it does it in the other order.

If we go back to [my original comment](https://github.com/Rapptz/discord.py/discussions/8263#discussioncomment-3212360), the first solution of adding an empty `@guilds` decorator and not needing the `remove_command` call at all does the trick, you can use this in your cog.

**Reply by Kadethra** [reply]
Thx a lot!
